ReactNativeClient/lib/joplin-database.js

```diff
@@ -686,6 +686,24 @@ class JoplinDatabase extends Database {
 				queries.push('ALTER TABLE version ADD COLUMN table_fields_version INT NOT NULL DEFAULT 0');
 			}
 
+			if (targetVersion == 30) {
+				queries.push('ALTER TABLE tags ADD COLUMN parent_id TEXT NOT NULL DEFAULT ""');
+				// Modify the tags_with_note_count view table for hierarchical tags
+				queries.push('DROP VIEW tags_with_note_count');
+				queries.push(`CREATE VIEW tags_with_note_count AS
+									WITH RECURSIVE
+									parent_of(id, parent_id, title, created_time, updated_time, child_id) AS 
+									(SELECT id, parent_id, title, created_time, updated_time, id FROM tags
+										UNION ALL
+										SELECT parent_of.id, parent_of.parent_id, parent_of.title, parent_of.created_time, parent_of.updated_time, tags2.id FROM parent_of JOIN tags AS tags2 ON parent_of.child_id=tags2.parent_id)
+									SELECT parent_of.id, parent_of.parent_id, parent_of.title, parent_of.created_time, parent_of.updated_time, COUNT(DISTINCT notes.id) as note_count FROM
+											parent_of
+											LEFT JOIN note_tags nt on nt.tag_id = parent_of.child_id 
+											LEFT JOIN notes on notes.id = nt.note_id 
+											WHERE notes.id IS NOT NULL
+										GROUP BY parent_of.id`);
+			}
+
```

Comment:
I'm not really keen on this big SQL query for the following reasons:
- It's huge and we don't want to encode too much business logic in the database
- Every time a field is added to the tag table, it should be duplicated here
- It adds a non-existing "note_count" field to the tag. It means if somewhere in the code something tries to save this tag it will fail. In general I've been quite strict about making sure that there's a one-to-one mapping between table fields and model fields so that you don't need to add extra code and checks before saving a model, and I'd prefer to keep it that way.

For folders we have a `Folder.addNoteCounts()` method, which I don't like much either, but at least it makes it clear from the calling code that we are somehow adding something to the list of folders.

But ideally there should be a separate structure, an associate array that maps a tag ID to a note count. And that structure would be updated whenever relevant. That way it's easy to access the note count, and the actual tag models are not modified.

I'm not sure how easy it is to implement though. Perhaps the note count could be removed from this pull request to keep things simple, and a second pull request could add that back.

Reply:
The reason I've kept the SQL query, was due to the way that note counting was done till now, which used a SQL query defined at `targetVersion == 25`. Obviously with recursion it became rather bulky, so I agree that it would be best to factor it out into its own model. I'll have a look into the effort required, or else to prevent the complication of this PR I will remove the note counting from this PR.

Reply:
Ok so for the moment I've removed the `tags_with_note_count` view table and implemented an equivalent function `Tag.addNoteCounts()` that counts the number of linked notes on the fly and adds the attribute to the object. Adding the function is a bit less overarching than (temporarily) removing the tag note count entirely. 

As to creating a separate structure for tag note count - it is also a bit more complex. The reason is that there is (at the moment) nothing stopping the user from tagging the same note with two tags from the same hierarchy e.g. `a/b` and `a/b/c`, in which case when incrementing the note count for a specific tag you would need to know if the note has already been associated with another tag from the same hierarchy. You could potentially solve this by maintaining a separate associative structure `tag_id->note_id`, which would link a tag with any of its descendant tags' notes.